Fix #648: allow using "old" (e.g. pre-3.0.4) plural rules evaluation

mauriziopinotti opened the pull request:
Add forcePluralCaseFallback option to force evaluation of fallback plural rules, i.e.
    
`forcePluralCaseFallback: false`
Default behavior, will use "zero" rule for 0 only if the language is set to do so (e.g. for "lt" but not for "en").
    
`forcePluralCaseFallback: true`
Force using "zero" rule for 0 even if the language doesn't use it by default (e.g. "en"). If "zero" localization for that string doesn't exist, "other" is still used as fallback.

mauriziopinotti committed:
Add .vscode/ and .idea/ folders to gitignore

mauriziopinotti committed:
Fix #648: allow using "old" (e.g. pre-3.0.4) plural rules evaluation

Add forcePluralCaseFallback option to force evaluation of fallback
plural rules, i.e.

* forcePluralCaseFallback: false
Default behavior, will use "zero" rule for 0 only if the language
is set to do so (e.g. for "lt" but not for "en").

* forcePluralCaseFallback: true
Force using "zero" rule for 0 even if the language doesn't use it
by default (e.g. "en"). If "zero" localization for that string
doesn't exist, "other" is still used as fallback.

bw-flagship commented:
Why would we allow modifying a language's plural behavior? What is the use case?

mauriziopinotti commented:
> Why would we allow modifying a language's plural behavior? What is the use case?

This is to restore the old behavior that made sense in a lot of use cases, where "zero" and "one" have better (more meaningful) variations than "other".

For example:

``` 
  "label_members": {
    "zero": "nobody",
    "one": "{} member",
    "two": "{} members",
    "other": "{} members"
  },
``` 

Here, "nobody" is better than "0 members" but without this fix I can't have it in English.
Also, this would create inconsistencies between English (displaying "0 members" and other languages such Latvian (displaying "nobody").

Another example:

```
  "label_attachments_count": {
    "zero": "No attachments",
    "one": "One single attachment",
    "two": "{} attachments",
    "other": "{} attachments"
  },
```

Here I have special versions of "zero" and "one" that would sound better for the user.

bw-flagship commented:
Thanks for the explanation!

I understand why that's helpful for your situation.

However, it feels wrong to do so because it missuses plurals.

Imho the correct solution would be:

```
  {
  "label_members_none": "nobody",
  "label_members": {
    "one": "{} member",
    "other": "{} members"
  },
```

Open for opinions, what do the others think?

mauriziopinotti commented:
> Thanks for the explanation!
> 
> I understand why that's helpful for your situation.
> 
> However, it feels wrong to do so because it missuses plurals.
> 
> Imho the correct solution would be:
> 
> ```
>   {
>   "label_members_none": "nobody",
>   "label_members": {
>     "one": "{} member",
>     "other": "{} members"
>   },
> ```
> 
> Open for opinions, what do the others think?

But the usage of this would be
```dart
final string = qty == 0 ? LocaleKeys.label_members_none.tr() : LocaleKeys.label_members_none.plural(qty);
```
which is a bit unreadable and it gets even worst with the second scenario:
```dart
final string = qty == 0 ? LocaleKeys.label_members_none.tr() : 
    qty == 1 ? LocaleKeys.label_members_one.tr() : 
        LocaleKeys.label_members_none.plural(qty);
```

Since the behavior is behind an optional flag I think it's good to let developers choose: if anyone disagrees they just have to do nothing, because out-of-the-box the behavior is the "correct" one.

Also notice that for the end user it's totally transparent: users will just see the desired text and they don't care how it's built under the hood.

bw-flagship commented:
Well, what is the scope of easy_localization?

Your goal is to show two entirely different texts based on the count of a variable.
"nobody" is not a plural form of "{} members", it is a different text.
Lets not abuse the plural engine for this.

I would not want to make this a feature of easy_localization.

mauriziopinotti commented:
> Well, what is the scope of easy_localization?
> 
> Your goal is to show two entirely different texts based on the count of a variable. "nobody" is not a plural form of "{} members", it is a different text. Lets not abuse the plural engine for this.
> 
> I would not want to make this a feature of easy_localization.

Well, the original issue has 5 thumb-ups so I'm clearly not the only one that feels this is a regression...

mauriziopinotti commented:
Furthermore, the official flutter packages behave in the same way, for example if I use
```json
{
  "pushed": "You have pushed the button this many times:",
  "nTimes": "{count, plural, =0{never} =1{once} other{{count} times}}"
}
```
```dart
Text(AppLocalizations.of(context)!.nTimes(_counter)),
```

Here's the output:

For "0":
![image](https://github.com/aissat/easy_localization/assets/4000539/6f9d2809-6b6d-4da3-8178-400e77dfcc9f)

For "1":
![image](https://github.com/aissat/easy_localization/assets/4000539/3bf07e99-ee04-49d0-9cde-dac9132769c5)

For "2":
![image](https://github.com/aissat/easy_localization/assets/4000539/513e5c09-2100-475b-b121-89d8d47d859e)

So, the previous behavior was legit.

Full example: https://github.com/easyhour/easy_localization/tree/i18n_demo

Official Flutter docs: https://docs.flutter.dev/ui/accessibility-and-internationalization/internationalization

bw-flagship commented:
Thanks for elaborating on this. The argument that the official flutter translations implementation has this behavior is very convincing.
One thing I would suggest is to rename the property and call it
`ignorePluralRules` instead of `forcePluralCaseFallback `

This would make the behavior more clear imho. Wdyt @mauriziopinotti ?

mauriziopinotti commented:
> Thanks for elaborating on this. The argument that the official flutter translations implementation has this behavior is very convincing. One thing I would suggest is to rename the property and call it `ignorePluralRules` instead of `forcePluralCaseFallback `
> 
> This would make the behavior more clear imho. Wdyt @mauriziopinotti ?

Sure, thanks, I'll submit an updated patch in the next days!

mauriziopinotti committed:
Rename forcePluralCaseFallback to ignorePluralRules and reverse logic

Parameter forcePluralCaseFallback has been renamed to ignorePluralRules
for clarity as suggested by bw-flagship. Also, the logic is now reversed
(e.g. false will implement the old behavior and true will implement the
new behavior). Finally, the default value is now false (old behavior).

mauriziopinotti commented:
Hi @bw-flagship , I've pushed a new commit with the renamed parameter.

While working on this I had two doubts
1. Should also the logic be reversed, i.e. before forcePluralCaseFallback=true is the modified behavior but now it seems that the modified behavior should be ignorePluralRules=false.
2. Should the default value be the pre-3.0.4 evaluation?

So, to clarify, now the patch works like this:
* **By default, the behavior is the _pre-3.0.4_ one**: "none, once, 2 times" (`ignorePluralRules=false`)
* Passing `ignorePluralRules=true` will cause the _after-3.0.5_ behavior: "0 times, 1 times, 2 times"

Let me know if this is OK or more changes are needed.

Thanks!

mauriziopinotti committed:
Fix unit tests

bw-flagship commented on lib/src/localization.dart:
```diff
@@ -114,6 +119,9 @@ class Localization {
   }
 
   static PluralRule? _pluralRule(String? locale, num howMany) {
+    if (!instance._ignorePluralRules) {
```
`instance._ignorePluralRules`

bw-flagship commented:
@mauriziopinotti I would reverse it:

ignorePluralRules is for me the same as ForceUsingFallback

ignorePluralRules = true should cause the behavior if 3.0.4 and earlier (and can be default)
ignorePluralRules = false should cause the "new" 3.0.5 behavior

mauriziopinotti committed:
Re-reverse logic of ignorePluralRules

ignorePluralRules=true, "old" beahvior (default)
ignorePluralRules=false, "new" beahvior

mauriziopinotti commented:
> @mauriziopinotti I would reverse it:
> 
> ignorePluralRules is for me the same as ForceUsingFallback
> 
> ignorePluralRules = true should cause the behavior if 3.0.4 and earlier (and can be default) ignorePluralRules = false should cause the "new" 3.0.5 behavior

OK, it should be good now.

bw-flagship commented on lib/src/localization.dart:
```diff
@@ -186,7 +194,8 @@ class Localization {
 
     final tag = '$key.$subKey';
-    var resource = _resolve(tag, logging: false, fallback: _fallbackTranslations != null);
+    var resource =
```
revert formatting change

bw-flagship commented on lib/src/localization.dart:
```diff
@@ -139,11 +147,11 @@ class Localization {
 
     final pluralRule = _pluralRule(_locale.languageCode, value);
-    final pluralCase = pluralRule != null ? pluralRule() : _pluralCaseFallback(value);
+    final pluralCase =
```
revert formatting change

bw-flagship commented on lib/src/localization.dart:
```diff
@@ -114,6 +119,9 @@ class Localization {
 
   static PluralRule? _pluralRule(String? locale, num howMany) {
+    if (instance._ignorePluralRules) {
+      return () => _pluralCaseFallback(howMany);
```
why not just

`return _pluralCaseFallback(howMany);`

mauriziopinotti replied:
`PluralRule` is defined as function `typedef PluralRule = PluralCase Function();` so I need to return a function with no args.

![image](https://github.com/aissat/easy_localization/assets/4000539/d64ec019-53be-494f-9e6f-650bc3af26bb)

bw-flagship commented:
the logic looks good! Found some formating things and you need to integrate the latest dev changes, thanks!

mauriziopinotti committed:
Revert some formatting changes

mauriziopinotti committed:
Merge remote-tracking branch 'upstream/develop' into fix_648
